prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1155 @@
+{
+	struct util_rx_entry *util_entry;
+
+	util_entry = util_get_recv_entry(srx,
+					 owner_entry->peer_entry.iov,
+					 owner_entry->peer_entry.desc,
+					 owner_entry->peer_entry.count, addr,
+					 owner_entry->peer_entry.context,
+					 owner_entry->peer_entry.tag,
+					 owner_entry->ignore,
+					 owner_entry->peer_entry.flags &
+					 (~FI_MULTI_RECV));
+	if (!util_entry)
+		return NULL;
+
+	if (util_adjust_multi_recv(srx, &owner_entry->peer_entry, size))
+		slist_remove_head(queue);
+
+	util_entry->peer_entry.owner_context = owner_entry;
+	owner_entry->multi_recv_ref++;
+
+	return util_entry;
+}
+
+static int util_match_msg(struct fid_peer_srx *srx, fi_addr_t addr, size_t size,
+			  struct fi_peer_rx_entry **rx_entry)
+{
+	struct util_srx_ctx *srx_ctx;
+	struct util_rx_entry *util_entry;
+	struct slist_entry *head;
+	int ret = FI_SUCCESS;
+
+	srx_ctx = srx->ep_fid.fid.context;
+	if (slist_empty(&srx_ctx->msg_queue)) {
+		util_entry = util_init_unexp(srx_ctx, addr, size, 0);
+		if (!util_entry)
+			return -FI_ENOMEM;
+		ret = -FI_ENOENT;
+	} else {
+		head = srx_ctx->msg_queue.head;
+		util_entry = container_of(head, struct util_rx_entry,
+					  peer_entry);
+		if (util_entry->peer_entry.flags & FI_MULTI_RECV) {
+			util_entry = util_process_multi_recv(srx_ctx,
+				&srx_ctx->msg_queue, addr, size, util_entry);
+			if (!util_entry) {
+				FI_WARN(&core_prov, FI_LOG_EP_CTRL,
+					"cannot allocate multi receive "
+					"buffer\n");
+				return -FI_ENOMEM;
+			}
+		} else {
+			(void) slist_remove_head(&srx_ctx->msg_queue);
+		}
+		srx_ctx->update_func(srx_ctx, util_entry);
+	}
+	util_entry->peer_entry.srx = srx;
+	*rx_entry = &util_entry->peer_entry;
+	return ret;
+}
+
+static int util_get_msg(struct fid_peer_srx *srx, fi_addr_t addr,
+		        size_t size, struct fi_peer_rx_entry **rx_entry)
+{
+	struct util_srx_ctx *srx_ctx;
+	struct util_rx_entry *util_entry, *any_entry;
+	struct slist *queue;
+	int ret;
+
+	srx_ctx = srx->ep_fid.fid.context;
+	srx_ctx->lock_acquire(&srx_ctx->lock);
+
+	queue = addr == FI_ADDR_UNSPEC ? NULL:
+		ofi_array_at(&srx_ctx->src_recv_queues, addr);
+
+	if (!queue || slist_empty(queue)) {
+		ret = util_match_msg(srx, addr, size, rx_entry);
+		goto out;
+	}
+
+	util_entry = container_of(queue->head, struct util_rx_entry,
+				  peer_entry);
+	if (!slist_empty(&srx_ctx->msg_queue)) {
+		any_entry = container_of(&srx_ctx->msg_queue.head,
+					 struct util_rx_entry, peer_entry);
+		if (any_entry->seq_no <= util_entry->seq_no) {
+			queue = &srx_ctx->msg_queue;
+			util_entry = any_entry;
+		}
+	}
+
+	if (util_entry->peer_entry.flags & FI_MULTI_RECV) {
+		util_entry = util_process_multi_recv(srx_ctx, queue, addr, size,
+						     util_entry);
+		if (!util_entry) {
+			FI_WARN(&core_prov, FI_LOG_EP_CTRL,
+				"cannot allocate multi receive buffer\n");
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+	} else {
+		(void) slist_remove_head(queue);
+	}
+
+	srx_ctx->update_func(srx_ctx, util_entry);
```

Comment:
Do you need to make this call after `util_entry->peer_entry.srx = srx;`? provider (efa) may use util_entry->peer_entry.srx inside update_func

Reply:
oh sure, good call. Thanks!

Reply:
Similarly for other places.

Reply:
@shijin-aws Thanks! I've fixed it locally. Was just waiting for the API PR to get merged before repushing to limit CI testing.